mlir/lib/Dialect/Rock/Transforms/GemmToGridwise.cpp

```diff
@@ -130,11 +137,183 @@ computeGridSizeAttentionGemmElmtGemm(ConversionPatternRewriter &rw, Op op,
+    // 1. addDim (gemmNSplit)
+    // 2. merge (gemmG, gemmNSplit) -> (gemmG*)
+    BottomUpTMBuilder addDimTransform(builder, {"gemmG", "gemmK", "gemmM"},
+                                      aShape, loc);
+
+    addDimTransform.passThrough({"gemmG", "gemmK", "gemmM"});
+    addDimTransform.addDim("gemmNSplit", 3, splitNFactor);
+    auto addDimTransformAttr = addDimTransform.get();
+
+    SmallVector<Attribute> transformAttrs;
+    transformAttrs.push_back(addDimTransformAttr);
+
+    auto mergeTransform =
+        BottomUpTMBuilder::above(addDimTransform, addDimTransformAttr);
+
+    mergeTransform.merge("gemmG", 0, {"gemmG", "gemmNSplit"});
+    mergeTransform.passThrough({"gemmK", "gemmM"});
+
+    auto mergeTransformAttr = mergeTransform.get();
+    transformAttrs.push_back(mergeTransformAttr);
+
+    std::reverse(transformAttrs.begin(), transformAttrs.end());
+    ArrayAttr arrayTransformAttrs = builder.getArrayAttr(transformAttrs);
+    aNew = mlir::rock::transform(builder, a, arrayTransformAttrs);
+  }
+
+  {
+    // Prepare matrix out - i.e., (gemmG, gemmM, gemmO)
+    // Using top-down transformations
+    // 1. merge (gemmG * gemmNSplit, gemmM, gemmO) -> (gemmG, gemmNSplit, gemmM,
+    // gemmO)
+    // 2. ignore (gemmG, gemmNSplit, gemmM, gemmN) -> (gemmG, gemmM, gemmO)
+
+    const int64_t G = outShape[0];
+    const int64_t M = outShape[1];
+    const int64_t O = outShape[2];
+
+    TopDownTMBuilder mergenTransform(builder, {"gemmG", "gemmM", "gemmO"},
+                                     {G * splitNFactor, M, O});
+
+    mergenTransform.merge({"gemmG", "gemmNSplit"}, {0, 1}, "gemmG",
+                          {G, splitNFactor});
+    mergenTransform.passThrough({"gemmM", "gemmO"}, {2, 3}, {"gemmM", "gemmO"});
+    auto mergenTransformAttr = mergenTransform.get();
+
+    SmallVector<Attribute> transformAttrs;
+    transformAttrs.push_back(mergenTransformAttr);
+
+    TopDownTMBuilder ignoreTransform =
+        TopDownTMBuilder::below(mergenTransform, mergenTransformAttr);
+
+    ignoreTransform.ignore("gemmNSplit");
+    ignoreTransform.passThrough({"gemmG", "gemmM", "gemmO"}, {0, 1, 2},
+                                {"gemmG", "gemmM", "gemmO"});
+
+    TransformMapAttr ignoreTransformAttr = ignoreTransform.get();
+    transformAttrs.push_back(ignoreTransformAttr);
+
+    ArrayAttr arrayTransformAttrs = builder.getArrayAttr(transformAttrs);
+    outNew = mlir::rock::transform(builder, out, arrayTransformAttrs);
+  }
+  return std::make_tuple(aNew, bNew, cNew, outNew);
+}
+
 static LogicalResult commonAttentionGemmElmtGemm(
     ConversionPatternRewriter &rw, RockGemmGemmWrapperInterface op, Value a,
     Value b, Value c, Value out, Value lse, Value currentSeqLen,
     UnitAttr causal, ValueRange elementwiseInputs, Region &preSecondOpRegion,
-    bool enableSoftmax, TypeAttr softmaxType) {
+    bool enableSoftmax, TypeAttr softmaxType,
+    const BufferDependencyAnalysis *bufferDeps) {
```

Comment:
Use const-ref. const pointer doesn't gurantee that underlying object is being modified by some other pointer to it. 
```suggestion
    const BufferDependencyAnalysis& bufferDeps) {
```

Reply:
I'm using a pointer because for attention this is nullptr.

Reply:
You can try using `std::optional`.